src/services/LogFileManager/index.ts

```diff
@@ -330,27 +339,42 @@ class LogFileManager {
             null !== this.#decoder.getFilteredLogEventMap()
         );
 
-        decodedEvents?.forEach(([message, , , logEventNum]) => {
+        if (null === decodedEvents) {
+            return;
+        }
+
+        for (const [message, , , logEventNum] of decodedEvents) {
             const matchResult = message.match(queryRegex);
-            if (null !== matchResult && "number" === typeof matchResult.index) {
-                const pageNum = Math.ceil(logEventNum / this.#pageSize);
-                if (false === results.has(pageNum)) {
-                    results.set(pageNum, []);
-                }
-                results.get(pageNum)?.push({
-                    logEventNum: logEventNum,
-                    message: message,
-                    matchRange: [
-                        matchResult.index,
-                        (matchResult.index + matchResult[0].length),
-                    ],
-                });
+            if (null === matchResult || "number" !== typeof matchResult.index) {
+                continue;
             }
-        });
+            const pageNum = Math.ceil(logEventNum / this.#pageSize);
+            if (false === results.has(pageNum)) {
+                results.set(pageNum, []);
+            }
+            results.get(pageNum)?.push({
+                logEventNum: logEventNum,
+                message: message,
+                matchRange: [
+                    matchResult.index,
+                    (matchResult.index + matchResult[0].length),
+                ],
+            });
 
-        this.#onQueryResults(results);
+            this.#queryCount++;
+            if (MAX_QUERY_RESULT_COUNT <= this.#queryCount) {
+                break;
+            }
+        }
+
+        // The query progress takes the maximum of the progress based on the number of events
+        // queried over total log events, and the number of results over the maximum result limit.
+        this.#onQueryResults(
+            Math.max(chunkEndIdx / this.#numEvents, this.#queryCount / MAX_QUERY_RESULT_COUNT),
+            results
+        );
```

Comment:
I agree with the rabit:
```suggestion
        // The query progress takes the maximum of the progress based on the number of events
        // queried over total log events, and the number of results over the maximum result limit.
        const progress = Math.max(
            chunkEndIdx / this.#numEvents,
            this.#queryCount / MAX_QUERY_RESULT_COUNT
        );

        this.#onQueryResults(progress, results);
```